gnocchi/incoming/ceph.py

```diff
@@ -190,40 +190,44 @@ def has_unprocessed(self, metric_id):
         with rados.ReadOpCtx() as op:
-            omaps, ret = self.ioctx.get_omap_vals(op, "", key_prefix, -1)
-            self.ioctx.operate_read_op(op, self.get_sack_name(sack),
-                                       flag=self.OMAP_READ_FLAGS)
-            # NOTE(sileht): after reading the libradospy, I'm
-            # not sure that ret will have the correct value
-            # get_omap_vals transforms the C int to python int
-            # before operate_read_op is called, I dunno if the int
-            # content is copied during this transformation or if
-            # this is a pointer to the C int, I think it's copied...
-            try:
-                ceph.errno_to_exception(ret)
-            except rados.ObjectNotFound:
-                # Object has been deleted, so this is just a stalled entry
-                # in the OMAP listing, ignore
-                return
-
-        measures = self._make_measures_array()
-        for k, v in omaps:
-            measures = numpy.concatenate(
-                (measures, self._unserialize_measures(k, v)))
-            processed_keys.append(k)
+            for metric_id in metric_ids:
+                sack = self.sack_for_metric(metric_id)
+                key_prefix = self.MEASURE_PREFIX + "_" + str(metric_id)
+                omaps, ret = self.ioctx.get_omap_vals(op, "", key_prefix, -1)
+                self.ioctx.operate_read_op(op, self.get_sack_name(sack),
+                                           flag=self.OMAP_READ_FLAGS)
+                # NOTE(sileht): after reading the libradospy, I'm
+                # not sure that ret will have the correct value
+                # get_omap_vals transforms the C int to python int
+                # before operate_read_op is called, I dunno if the int
+                # content is copied during this transformation or if
+                # this is a pointer to the C int, I think it's copied...
+                try:
+                    ceph.errno_to_exception(ret)
+                except rados.ObjectNotFound:
+                    # Object has been deleted, so this is just a stalled entry
+                    # in the OMAP listing, ignore
+                    continue
+
+                m = self._make_measures_array()
+                for k, v in omaps:
+                    m = numpy.concatenate(
+                        (m, self._unserialize_measures(k, v)))
+                    processed_keys[sack].append(k)
+
+                measures[metric_id] = m
 
         yield measures
 
         # Now clean omap
         with rados.WriteOpCtx() as op:
-            # NOTE(sileht): come on Ceph, no return code
-            # for this operation ?!!
-            self.ioctx.remove_omap_keys(op, tuple(processed_keys))
-            self.ioctx.operate_write_op(op, self.get_sack_name(sack),
-                                        flags=self.OMAP_WRITE_FLAGS)
+            for sack, keys in six.iteritems(processed_keys):
```

Comment:
i'm not sure we need to group things by sack here. iiuc, this command only gets triggered by sack, never cross-sacks  as it only ever locks one sack.

Reply:
This is not what the signature of the function says, so why your right in theory, it would be wrong to do that.

Reply:
*shrugs* i meant to pass it in but i guess you can do it as a followup.

Reply:
It's in the roadmap :)